File: linked-list-cycle-ii/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {"}": "{", "]": "[", ")": "("}
        
        for char in s:
            if char in mapping:
                popped = stack.pop() if stack else "."
                if (popped != mapping[char]):
                    return False
            else:
                stack.append(char)
        return not stack

File: valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> um = {{')', '('}, {'}', '{'}, {']', '['}};
        stack<char> st;
        for (char c : s) {
            if (um.find(c) != um.end()) {
                char p = !st.empty() ? st.top() : '.';
                if (!st.empty()) st.pop();
                if (p != um[c]) {
                    return false;
                }
            } else {
                st.push(c);
            }
        }
        return st.empty();
    }
};
